File: seonize-backend/app/schemas/topical_map.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime

class TopicalKeywordBase(BaseModel):
    keyword: str
    search_volume: int = 0
    cpc: float = 0.0
    competition: float = 0.0
    intent: Optional[str] = None
    suggested_title: Optional[str] = None

class TopicalKeywordResponse(TopicalKeywordBase):
    id: int
    cluster_id: str
    status: str
    created_at: datetime

    class Config:
        from_attributes = True

class TopicalClusterBase(BaseModel):
    name: str
    description: Optional[str] = None
    level: int = 1

class TopicalClusterResponse(TopicalClusterBase):
    id: str
    topical_map_id: str
    parent_id: Optional[str] = None
    created_at: datetime
    keywords: List[TopicalKeywordResponse] = []
    subclusters: List['TopicalClusterResponse'] = []

    class Config:
        from_attributes = True

class TopicalMapBase(BaseModel):
    name: str
    topic: str
    country: str = "TW"
    language: str = "zh-TW"

class TopicalMapCreate(TopicalMapBase):
    pass

class TopicalMapResponse(TopicalMapBase):
    id: str
    user_id: str
    total_keywords: int
    total_search_volume: int
    status: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

# For the tree view visualization
class TopicalMapDetailResponse(TopicalMapResponse):
    clusters: List[TopicalClusterResponse] = []

# To update the Pydantic self-referencing model
TopicalClusterResponse.model_rebuild()

File: seonize-frontend/src/types/topical-map.ts
export interface TopicalKeyword {
  id: number;
  cluster_id: string;
  keyword: string;
  search_volume: number;
  cpc: number;
  competition: number;
  intent?: string;
  suggested_title?: string;
  status: string;
  created_at: string;
}

export interface TopicalCluster {
  id: string;
  topical_map_id: string;
  parent_id?: string;
  name: string;
  description?: string;
  level: number;
  created_at: string;
  keywords: TopicalKeyword[];
  subclusters: TopicalCluster[];
}

export interface TopicalMap {
  id: string;
  user_id: string;
  name: string;
  topic: string;
  country: string;
  language: string;
  total_keywords: number;
  total_search_volume: number;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  created_at: string;
  updated_at: string;
}

export interface TopicalMapDetail extends TopicalMap {
  clusters: TopicalCluster[];
}

export interface CreateTopicalMapRequest {
  name: string;
  topic: string;
  country?: string;
  language?: string;
}
